core/ia32.c

```diff
@@ -0,0 +1,196 @@
+/*
+* Copyright (c) 2011 Intel Corporation
+*
+* Redistribution and use in source and binary forms, with or without
+* modification, are permitted provided that the following conditions are met:
+*
+*   1. Redistributions of source code must retain the above copyright notice,
+*      this list of conditions and the following disclaimer.
+*
+*   2. Redistributions in binary form must reproduce the above copyright
+*      notice, this list of conditions and the following disclaimer in the
+*      documentation and/or other materials provided with the distribution.
+*
+*   3. Neither the name of the copyright holder nor the names of its
+*      contributors may be used to endorse or promote products derived from
+*      this software without specific prior written permission.
+*
+* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+* AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+* IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+* ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+* LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+* CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+* SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+* INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+* CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+* ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+* POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#include "../include/hax.h"
+#include "../include/asm.h"
+
+struct qword_val {
+    uint32 low;
+    uint32 high;
+};
+
+extern void ASMCALL asm_enable_irq(void);
+extern void ASMCALL asm_disable_irq(void);
+
+extern void asm_btr(uint8 *addr, uint bit);
+extern void asm_bts(uint8 *addr, uint bit);
+
+extern void asm_fxinit(void);
+extern void asm_fxsave(mword *addr);
+extern void asm_fxrstor(mword *addr);
+
+
+#ifdef _M_IX86
+extern uint64 ASMCALL asm_vmread(uint32 component);
```

Comment:
BTW, the 64-bit version of `asm_vmread` should also return `mword` instead of `uint32`. According to @junxiaoc , this change alone is enough to make this patch work on Windows.